avocado_vt/plugins/vt_joblock.py

```diff
@@ -0,0 +1,120 @@
+                                            key="show_lock_location",
+                                            key_type=bool,
+                                            default=False)
+        self.lock_file = None
+
+    def _abort(self, message, job):
+        """
+        Aborts the Job by exiting Avocado, adding a failure to the job status
+        """
+        self.log.error(message)
+        sys.exit(exit_codes.AVOCADO_JOB_FAIL | job.exitcode)
+
+    def _set_lock(self, job):
+        if not os.path.isdir(self.lock_dir):
+            msg = ('VT Job lock directory "%s" does not exist... '
+                   'exiting...' % self.lock_dir)
+            self._abort(msg, job)
+
+        pattern = 'avocado-vt-joblock-%(jobid)s-%(uid)s-%(random)s.pid'
+        # the job unique id is already random, but, let's add yet another one
+        rand = ''.join([random.choice(string.ascii_lowercase + string.digits)
+                        for i in xrange(8)])
+        path = pattern % {'jobid': job.unique_id,
+                          'uid': os.getuid(),
+                          'random': rand}
+        path = os.path.join(self.lock_dir, path)
+
+        try:
+            with open(path, 'w') as lockfile:
+                lockfile.write("%u" % os.getpid())
+            self.lock_file = path
+        except IOError as e:
+            msg = ('Failed to create VT Job lock file. Exiting...')
+            self._abort(msg, job)
+
+        if self.show_lock:
+            self.log.info("VT LOCK    : %s", self.lock_file)
+
+    def _get_lock_file(self):
+        if not os.path.isdir(self.lock_dir):
+            return None
+
+        try:
+            files = os.listdir(self.lock_dir)
+            if not files:
+                return None
+        except OSError as e:
+            if e.errno == errno.ENOENT:
+                return None
+
+        pattern = re.compile(r'avocado-vt-joblock-[0-9a-f]{40}-[0-9]+'
+                             '-[0-9a-z]{8}\.pid')
+        for lock_file in files:
+            if pattern.match(lock_file):
+                path = os.path.join(self.lock_dir, lock_file)
+                if os.path.isfile(path):
+                    content = int(open(path, 'r').read())
+                    if int(content) > 0:
+                        return path
+        return None
+
+    def _get_lock_pid(self):
+        lockfile = self._get_lock_file()
+        if lockfile is None:
+            return 0
+        return int(open(lockfile, 'r').read())
+
+    def _lock(self, job):
+        lock_pid = self._get_lock_pid()
+        if lock_pid > 0:
+            msg = ("Avocado-VT job lock acquired by PID %u. "
+                   "Aborting..." % lock_pid)
+            self._abort(msg, job)
+        self._set_lock(job)
+
+    def _unlock(self):
+        if self.lock_file:
+            os.unlink(self.lock_file)
+
+    def pre(self, job):
+        suite = job._make_test_suite(job.args.url)
```

Comment:
Don't take me wrong, I'd prefer having plugin to do the dirty work. I'm not happy about using plugin, which was just introduced, still a bit under discussion and most importantly which does not scale to implement this properly. Because of that you had to implement really nasty workaround which uses private methods of different class, which additionally is part of an internal avocado API.

Problem number two is, that the called method is really power-hungry; `avocado run boot --dry-run` takes 1.5s, but normally it finishes under a second (using 5 boot tests takes 5.5s vs. 3.3s). I know we're not interested in optimization at first, but it shows that something is not used as should.

I have nothing against JobPre/Post plugin, I think we ought to add even more hooks and only by using avocado we discover what are the most appropriate places. In the JobPre PR I mentioned, that it's hard to say when the JobPre should take place, this PR demonstrates, that some might be interested in other information. Maybe we should pass the `Job` itself and allow the 2 hooks (init and pre), then this could only look at the Job and evaluate if avocado-vt is used.

Alternatively I provided a simple modification to avocado-vt loader, which could solve this effectively, even without going through results (yes, the demonstration is incomplete, but I guess it can fit under 30 lines including multiple url evaluating and the multi-pid handling copied from your solution). Which brings a question, whether more plugins might be interested in this. Maybe it's another place for a hook. (I don't think so, I think better solution would be to allow loader plugins to mark the job with custom tag and the init/pre/post plugin should be able to query for the tag).

But this discussion is for later, first we need to see what is the most appropriate place to solve the issue and then we should modify `avocado` to support it nicely. Additionally as this is considered temporary workaround, we should consider if the permanent solution won't eliminate the need for such hook, but I think that even with locking API we need the entry point and the avocado-vt loader seems like the most appropriate place for it.

About the direction I think avocado should be kept as stable as possible. We should first think what are the requirements and if they are valid. Then and only then we should consider the most appropriate generic solution and implement it inside avocado.

For linked projects we need to think what's the best for the project, not for avocado. We have to raise questions and use cases. If something looks like not-really-useful for others, we should try implementing it in the project, rather than pushing it to avocado. This does not mean we should not reevaluate the decision once it becomes stable and we are certain others might benefit from it too. Then we should ask again and the avocado team should reconsider and try to find the generic API which the linked projects can use.

So the example here is, avocado-vt needs:

1. was certain loader used?
2. if it was, can I register function on job exit?

Right now avocado can't provide such information. As it's a temporary hack, we should just say, "Sorry guys, do it inside your project for now". Later we can revisit this decision and say, OK we have a time, so you need to know if loader was used? Fine we don't want to follow your solution, but we can offer you this instead, would it suit you? ...